### PR TITLE
Restart MicroShift when IP address changes

### LINKED ISSUE #556: [BUG] If the interface IP changes Microshift stops working
#### What happened:

The openshift-api api endpoints are created based on the existing IP address when microshift starts,
if the IP changes, the kubernetes controller can't connect to the endpoint, and starts failing.

#### What you expected to happen:
Microshift should have recognized the IP change and updated the endpoints, or used a VIP for the API purposes.

#### How to reproduce it (as minimally and precisely as possible):

1. Assign an IP address to the MicroShift host
2. Start MicroShift
3. Change the IP address
4. MicroShift stops working
5. restart MicroShift
6. MicroShift still not working

### PR DESCRIPTION
**Pull request by oglok**
MicroShift endpoints rely on the host IP address. If that IP changes for whatever reason, APIs will be completely unavailable and MicroShift will stop working. This PR will detect IP host changes, and shutdown MicroShift. Since the production deployment model is based on systemd service, MicroShift service will be automatically restarted and endpoints will be recreated to be back in a functional state. 

Signed-off-by: Miguel Angel Ajo <majopela@redhat.com>
Signed-off-by: Ricardo Noriega <rnoriega@redhat.com>

**Which issue(s) this PR addresses**:

Closes #556

### CODE REVIEW

**Comment by oglok** [comment]
To show some testing. After the IP address changes, it is detected and MicroShift exits:

```
W0408 13:17:35.906386  595827 ipwatch.go:58] IP address has changed from "192.168.0.140" to "192.168.0.99", restarting MicroShift                                                
```

Systemd service will restart it, and endpoints will change:

```
I0408 13:22:29.958411  599949 apiservice.go:86] deleting outdated endpoints openshift-apiserver
I0408 13:22:29.997946  599949 apiservice.go:86] deleting outdated endpoints openshift-oauth-apiserver
```

File: pkg/controllers/apiservice.go [hunk 1]
```diff
@@ -44,6 +46,56 @@ func createAPIHeadlessSvc(cfg *config.MicroshiftConfig, svcName string, svcPort
...
 
 	client := coreclientv1.NewForConfigOrDie(rest.AddUserAgent(restConfig, "core-agent"))
+
+	if err = ensureService(client, svcName); err != nil {
+		return errors.Wrapf(err, "Error creating service %q", svcName)
+	}
+
+	if err != ensureIPEndpoints(client, svcName, cfg.NodeIP, svcPort) {
+		return errors.Wrapf(err, "Error creating IP endpoints for service %q, IP %q:%d", svcName, cfg.NodeIP)
+	}
```
Comment (mangelajo) [inline pkg/controllers/apiservice.go:56]: ^checker complains;, the port is missing in the args?

File: pkg/controllers/apiservice.go [hunk 2]
```diff
@@ -64,52 +116,18 @@ func createAPIHeadlessSvc(cfg *config.MicroshiftConfig, svcName string, svcPort
 			},
 		},
 	}
-	_, err = client.Services("default").Get(context.TODO(), svc.Name, metav1.GetOptions{})
+
+	_, err := client.Services("default").Get(context.TODO(), svc.Name, metav1.GetOptions{})
 	if apierrors.IsNotFound(err) {
-		klog.Infof("Creating service %s", svc.Name)
+		logrus.Infof("creating svc %s", svc.Name)
```
Comment (mangelajo) [inline pkg/controllers/apiservice.go:122]: why logrus? no idea if it was originally in my patch, could be: bad Ajo! X'D

**Review by mangelajo — COMMENTED** [review]
some comments, thanks for handling it

File: pkg/controllers/apiservice.go (hunk 1 above)
Comment (oglok) [reply pkg/controllers/apiservice.go:55]: Fixed! Thx

File: pkg/controllers/apiservice.go (hunk 2 above)
Comment (oglok) [reply pkg/controllers/apiservice.go]: My bad too, the original PR was before the klog replacement.

**Comment by mangelajo** [comment]
/retest

Sent with a Spark

**Comment by mangelajo** [comment]
Oh, it's not CI, something is broken: 

```
[root@fedora ~]# oc get endpoints -A
NAMESPACE           NAME                      ENDPOINTS                                      AGE
default             kubernetes                192.168.1.70:6443                              2m12s
openshift-dns       dns-default               10.42.0.4:5353,10.42.0.4:9154,10.42.0.4:5353   83s
openshift-ingress   router-external-default                                                  83s
openshift-ingress   router-internal-default                                                  83s
[root@fedora ~]# oc get pods -A
NAMESPACE                       NAME                                  READY   STATUS    RESTARTS   AGE
kube-system                     kube-flannel-ds-s7tbr                 1/1     Running   0          5m16s
kubevirt-hostpath-provisioner   kubevirt-hostpath-provisioner-xqsg7   1/1     Running   0          5m6s
openshift-dns                   dns-default-n4kl9                     2/2     Running   0          5m16s
openshift-dns                   node-resolver-srpnj                   1/1     Running   0          5m16s
openshift-ingress               router-default-85bcfdd948-nprmj       0/1     Running   2          5m21s
openshift-service-ca            service-ca-7764c85869-nnv2j           1/1     Running   0          5m21s
```

File: pkg/controllers/apiservice.go [hunk 3]
```diff
@@ -44,6 +45,55 @@ func createAPIHeadlessSvc(cfg *config.MicroshiftConfig, svcName string, svcPort
...
+	if endpoints != nil {
+		if !reflect.DeepEqual(endpoints.Subsets, expectedEndpoints.Subsets) {
+			klog.Infof("deleting outdated endpoints %s", endpoints.Name)
+			if err := client.Endpoints("default").Delete(context.TODO(), endpoints.Name, metav1.DeleteOptions{}); err != nil {
+				return errors.Wrapf(err, "Error deleting outdated endpoints %q", endpoints.Name)
+			}
+		}
+	}
+
+	if apierrors.IsNotFound(err) || endpoints != nil {
```
Comment (mangelajo) [inline pkg/controllers/apiservice.go:88]:
```suggestion
	if apierrors.IsNotFound(err) || endpoints != nil {
```

This logic is wrong ^initial creation will never happen, probably I didn't test my initial version well enough (no endpints).

**Comment by mangelajo** [comment]
gogogo %)

**Comment by oglok** [comment]
There was more subtleness into the code than I expected xD xD

DeepEqual condition was not working properly, so I decided to change a bit the logic. Check if that endpoints name exist, do a proper deepEqual check and return if it exists, otherwise, create the endpoints. 

Tests are ok locally, let's see CI.

**Review by mangelajo — APPROVED** [review]
/lgtm

**Pull request merged**